py/server/deephaven/table.py

```diff
@@ -62,6 +64,106 @@ def _j_py_script_session() -> _JPythonScriptSession:
+        ...
+    elif return_type is None:
+        raise ValueError("Missing target function return type.")
+    else:
+        raise ValueError("Missing target function parameter type(s)")
+
+
+def _chunked(parameter_types: List[np.dtype] = None, return_type: np.dtype = None):
+    """A decorator to wrap a Python function used in Deephaven query formulas and invoked on a row basis. The wrapper
+    function is able to accept arrays of the declared parameter data type (vs. scalar), thus minimize the Java/Python
+    boundary crossings.
+
+    Args:
+        parameter_types: ...
+        return_type: ...
+    """
+    np_type_codes = _analyze_func_signature(parameter_types=parameter_types, return_type=return_type)
+
+    def decorator(fn: Callable):
+        @wraps(fn)
+        def wrapper(*args):
+            if len(args) != len(parameter_types):
+                raise ValueError("The number of arguments doesn't match the definition.")
+
+            chunk_size = len(args[0])
+            chunk_result = np.empty(chunk_size, return_type)
+            i = 0
+            for scalar_args in zip(*args):
+                chunk_result[i] = fn(*scalar_args)
+                i += 1
+            return chunk_result
+
+        wrapper.dh_chunked = True
+        wrapper.signature = np_type_codes
+        return wrapper
+
+    return decorator
+
+
+def encode_signature(fn: Callable) -> str:
+    try:
+        sig = inspect.signature(fn)
+    except:
+        return ""
+
+    parameter_types = []
+    for n, p in sig.parameters.items():
+        try:
+            np_dtype = np.dtype(p.annotation if p.annotation else "object")
+            parameter_types.append(np_dtype)
+        except TypeError:
+            parameter_types.append(np.dtype("object"))
+
+    try:
+        return_type = np.dtype(sig.return_annotation if sig.return_annotation else "object")
+    except TypeError:
+        return_type = np.dtype("object")
+
+    np_type_codes = [np.dtype(p).char for p in parameter_types]
+    np_type_codes.extend(["-", ">", np.dtype(return_type).char])
+    return "".join(np_type_codes)
+
+
+def dh_vectorized(fn: Callable):
+
+    sig_code = encode_signature(fn)
+    return_type = sig_code[-1]
+
+    def wrapper(*args):
+        if len(args) != len(sig_code) - len("->?"):
+            raise ValueError(f"The number of arguments doesn't match the definition. {len(args)}, {sig_code}")
+
+        if not args:
+            chunk_size = 4096
+        else:
+            chunk_size = len(args[0])
+        chunk_result = np.empty(chunk_size, return_type)
+
+        i = 0
+        if not args:
+            for i in range(chunk_size):
+                chunk_result[i] = fn()
+        else:
+            for scalar_args in zip(*args):
+                chunk_result[i] = fn(*scalar_args)
+                i += 1
+        return chunk_result
+
+    wrapper.dh_vectorized = True
+    wrapper.signature = sig_code
+
+    return wrapper
```

Comment:
A question about how this works in practice.  Will the following work?  Basically can users still call the functions in a scalar way, or does the wrapping break that?

```python
@dh_vectorized
def f(x):
    return x + 1
print(f(1))
t = t.update("Y=f(X)")
```

Reply:
Yes, this works too.